### PR TITLE
Don't swallow errors coming from the shareSession call

### PR DESCRIPTION
**Pull request by poljar**
This PR intends to fix https://github.com/vector-im/element-web/issues/23792.

What should have been an easy fix sadly turned out to be a deciphering of the logic here. It seems to me that the `ensureOutboundSession()` has a bug which would wedge the outbound group session in the case of an error.

This wasn't problematic until now, since almost no errors happen in the `prepareSession()` call, but as we found out may happen in the `shareSession` call.

The first patch fixes the mentioned bug, aligning the code with the comments.
The second patch now propagates errors from the `shareSession` call to the caller.

The individual commits have even more info around what each of them does and means.

I tested this in Element Web, if there's an error in `shareSession` we get prompted to retry the sending of the event. If the error happened to be transient, a press of the retry button will successfully share the outbound session and send the event.

This closes:  https://github.com/vector-im/element-web/issues/23792.

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Don't swallow errors coming from the shareSession call ([\#2908](https://github.com/matrix-org/matrix-js-sdk/pull/2908)). Fixes vector-im/element-web#23792. Contributed by @poljar.

### CODE REVIEW

File: src/crypto/algorithms/megolm.ts
Comment (richvdh) [inline src/crypto/algorithms/megolm.ts:281]: it seems like this is an implementation detail which belongs in a comment inside the function, rather than in the doc-comment?

File: src/crypto/algorithms/megolm.ts
Comment (richvdh) [inline src/crypto/algorithms/megolm.ts:305]: Given that the error will be logged in the `catch` handler later, this seems redundant.

File: src/crypto/algorithms/megolm.ts
Comment (richvdh) [inline src/crypto/algorithms/megolm.ts:313]: ... and make sure that the promise chain remains unbroken.

File: src/crypto/algorithms/megolm.ts
Comment (richvdh) [inline src/crypto/algorithms/megolm.ts:321]: I'd be tempted to inline `infallible` here.

**Review by richvdh — CHANGES_REQUESTED** [review]

File: src/crypto/algorithms/megolm.ts
Comment (poljar) [reply src/crypto/algorithms/megolm.ts:281]: Considering that this method is private, and modifies a field of the class, I would say that it's perfectly reasonable to explain how it does so in the doc comment.

Comment (richvdh) [reply src/crypto/algorithms/megolm.ts:281]: Not entirely convinced, but I don't feel that strongly.

File: src/crypto/algorithms/megolm.ts
Comment (poljar) [reply src/crypto/algorithms/megolm.ts]: Removed in https://github.com/matrix-org/matrix-js-sdk/pull/2908/commits/d5074776f117afe2af16756bfb74a3c9063740da.

**Review by richvdh — APPROVED** [review]
lgtm

**Comment by poljar** [comment]
Force pushed to get rid of the fixup commits.

**Comment by poljar** [comment]
Force pushed to resolve a merge conflict.

What is needed to get this merged? I don't have write access in this repo, if the expectations is that I'm going to do that.

**Comment by richvdh** [comment]
Oh, sorry, I didn't realise you didn't have write access. And now we have lint failures :(.

> Force pushed to resolve a merge conflict.

It should be possible to resolve a conflict without force-pushing?

**Comment by poljar** [comment]
> It should be possible to resolve a conflict without force-pushing?

Sure, but I posted some fixup commits that required a rebase/autosquash workflow so I continued it. Are rebases frowned upon in this repo, should I stop using them as well as fixup commits?

**Comment by richvdh** [comment]
Rebases after a review make it hard to see what has been changed since the review, so one tends to have to do the review all over again.

**Comment by poljar** [comment]
> Rebases after a review make it hard to see what has been changed since the review, so one tends to have to do the review all over again.

I understand this, that's why the fixup commits were used before the approval. Are you re-reviewing my merge-conflict resolution? Should I not use fixup commits for this lint issue now?

**Comment by richvdh** [comment]
@poljar unfortunately this is failing the test-coverage gate :(

**Comment by poljar** [comment]
This has now a unrelated change that I'm trying to merge with https://github.com/matrix-org/matrix-js-sdk/pull/2959.

I'm closing this and going to reopen two separate PRs, one for the fix for `setupPromise` becoming wedged, and one to stop swallowing the errors.

**Pull request closed without merging**
